components/scream/src/physics/share/common_physics_impl.hpp

```diff
@@ -0,0 +1,165 @@
+#ifndef COMMON_PHYSICS_IMPL_HPP
+#define COMMON_PHYSICS_IMPL_HPP
+
+#include "common_physics_functions.hpp"
+#include "physics_constants.hpp"
+
+namespace scream {
+namespace physics {
+
+  
+//-----------------------------------------------------------------------------------------------//
+// Applies Exners Function which follows:
+//   Exner = (P/P0)^(Rd/Cp),
+// where,
+//   P  is the pressure at this location, Pa
+//   P0 is a reference pressure, Pa
+//   Rd is the gas constant, J/K
+//   Cp is heat capacity of dry air, J/Ki
+// All universal constants, P0, Rd, and Cp, are defined in physics_constants.hpp
+// Note: Another experssion for Exner is,
+//   Exner = T/th
+// whre,
+//   T  is the temperature, K
+//   th is the potential temperature, K
+template<typename DeviceT>
+template<typename ScalarT>
+KOKKOS_FUNCTION
+ScalarT PhysicsFunctions<DeviceT>::get_exner(const ScalarT& p_mid)
+{
+  using C = scream::physics::Constants<ScalarT>;
+  static constexpr ScalarT p0(C::P0);
+  static constexpr ScalarT rd(C::RD);
+  static constexpr ScalarT inv_cp(C::INV_CP);
+
+  ScalarT exner = pow( p_mid/p0, rd*inv_cp );
+  // Simple check to make sure no obvious errors occurred
+  EKAT_REQUIRE_MSG(!isnan(exner),"Error in get_exner, exner(p = " + std::to_string(p_mid) + ") = NaN"); 
+  EKAT_REQUIRE_MSG(!(exner<0),   "Error in get_exner, exner(p = " + std::to_string(p_mid) + ") = " + std::to_string(exner) + " < 0");
+  // Return exner 
+  return exner;
+}
+// For operations on a full column at a time:
+template<typename DeviceT>
+template<typename ScalarT, typename InputProviderP>
+KOKKOS_FUNCTION
+void PhysicsFunctions<DeviceT>::get_exner(const MemberType& team,
+                      const InputProviderP& p_mid, 
+                      const view_1d<ScalarT>& exner)
+{
+  Kokkos::parallel_for(Kokkos::TeamThreadRange(team,exner.extent(0)),
+                       [&] (const int k) {
+    exner(k) = get_exner(p_mid(k)); 
+  });
+}
+
+//-----------------------------------------------------------------------------------------------//
+// Converts temperature to potential temperature using Exners function:
+//   theta = T_atm/exner(p_mid),
+// where
+//   theta  is the potential temperature, K
+//   T_mid  is the temperature, K
+//   p_mid  is the pressure, Pa.  Used for exners formula, see definition above, unitless
+template<typename DeviceT>
+template<typename ScalarT>
+KOKKOS_FUNCTION
+ScalarT PhysicsFunctions<DeviceT>::get_theta_from_T(const ScalarT& T_mid, const ScalarT& p_mid)
+{
+  ScalarT exner = get_exner(p_mid);
+  ScalarT theta = T_mid/exner;
+  // Simple check to make sure no obvious errors occurred
+  EKAT_REQUIRE_MSG(!isnan(theta),"Error in get_theta_from_T, theta(T = " + std::to_string(T_mid) + ", p = " + std::to_string(p_mid) + ") = NaN"); 
```

Comment:
Here too: I find more insightful to check inputs than outputs.

Reply:
Also, `to_string` is not overloaded for packs.

Reply:
In response to checking inputs rather than outputs.  My reasoning was that it would be helpful to make sure that the function doesn't return a NaN even if passed a reasonable set of inputs.  Think for example if you had a function that accepted `qc` as input, which could be 0.0 but some adjustment to the code accidentally divided by `qc` rather than multiplied.  The reasonable input of 0.0 would produce NaN as output which is precisely the type of thing we want to catch right?

Reply:
But checking both input and output should be simple and satisfy both concerns

Reply:
> Also, `to_string` is not overloaded for packs.

Is this something we can overload in EKAT?  Would we even want to?  I guess in this case it isn't essential to provide all of the information I am adding into the ERROR message.

Reply:
I don't think this fcn should bother about output. As long as inputs are valid, the fcn is good to go. If the user has constraints on the output values, _the user_ should check those. The checks on the inputs are to verify that it is "legal" to call the fcn. And most likely they are enough to guarantee a valid output. But again, what "valid" means for the output, depends on how it is used. And in general, that depends on how it used later.

Reply:
I agree with Luca about checking input only within the implementation of a function. However, I do think we should check function output in unit tests for the function--that's really where such checks belong, and will allow us to track bugs and fix them. To me, checking the output of a function given valid input blurs the boundary between model logic and tests. I do agree with the sentiment that we need to check output somewhere, though!

Reply:
@jeff-cohere , yes.  I went with this suggestion for the latest version.  But it would be great for you to keep an eye on the route I took in case there is an even better way.  At the moment there are no checks inside the functions themselves.  Instead, the unit tests do all the typical checking of the function output.